### ISSUE TITLE
a feature request for batch tagging

### ISSUE THREAD

**Issue by rockonedege** [sandreas/tone#42]
I wish to have this workflow

1. dump a folder(optionally recursively) of audios into a `json` file and 
2. edit the `json` file with an editor or script
3. import the `json` to update the original files

Is this already possible?

**Comment by sandreas** [comment]
Hey @rockonedege,

this is a good feature request. Unfortunately the `dump` part is not possible out of the box at the moment, but I will definitely give this a go for the next release `0.1.6`. It is important to have such a thing.

- [ ] Add `--export-file` (and optionally `--export-file-template`) parameters for `dump` command (e.g. `--export-file` and `--export-file-template="{filename}.tone.json"`)
  - `--export-file` is boolean, will switch the default format to `tonejson` and uses the `--export-file-template` as current file to dump with the valid extension
  - `--export-file-template` is for later and may influence the way exports can be done

If you are on Linux or macOS, you could try something like this for the dump:

```
find . -type f -name '*.m4b' -exec /bin/sh -c 'FULLPATH="{}";NOEXT="${FULLPATH%.m4b}" && tone dump "$FULLPATH" --format=json > "$NOEXT.tone.json"' \;
```

Basically this uses find and some shell kung fu to call `tone dump` for every `m4b` file and then exports it into `<filename>.tone.json`. Feel free to replace `m4b` with other extensions.

**Comment by rockonedege** [comment]
Thanks, I will give it try. 

Assuming it works as expect, what is the counterpart `tag` option to import the `json`  and apply it to audio files?

**Comment by sandreas** [comment]
tone tag audiobooks/ --auto-import toneJson

If i remember correctly

**Comment by rockonedege** [comment]
looking at the examples in the documents
```
tone tag --debug --auto-import=covers --meta-additional-field ©st3=testing input.m4b --dry-run
```
`--auto-import` seems expect a field name, not a json file? I am not sure I understand teh `auto-` part in the naming.

there's another option `--meta-tone-json-file ` seems more appropriate by name, but I fund no example.

**Comment by sandreas** [comment]
`--auto-import` is an Enumberable containing the following possible values (not filenames)

- Covers
- Chapters
- ToneJson
- Ffmetadata

Given one of these, `tone` will look for a specific file beneath to the location of the file(s) to tag, where `--meta-tone-json-file` means ONE specific file.

So `--auto-import=ToneJson` is made for recursive scanning and will take every <file>, look for an according `<filename-without-extension>.tone.json` and import it if it finds one.

Example:
- `Fantasy/Harry Potter/1/Harry Potter 1.m4b` => `Fantasy/Harry Potter/1/Harry Potter 1.tone.json`

Try it, it will do, what you want.

**Comment by rockonedege** [comment]
The `dump` command works perfectly and generates a `tone.json` file for each `.mp3`.
```find . -type f -name '*.mp3' -exec /bin/sh -c 'FULLPATH="{}";NOEXT="${FULLPATH%.mp3}" && tone dump "$FULLPATH" --format=json > "$NOEXT.tone.json"' \;
```

then I picked one of the tone json files. changed(cleared) some fields
```json
  "meta": {
    "album": "¸ü¶à×ÊÔ´Çë¼Ócqhdy2016",
    "artist": "Ò»ÊÖ×ÊÔ´¹Ù·½Í¬²½¸üÐÂÇë¼ÓÎ¢ÐÅ£ºcqhdy2016",
    "comment": "¸ü¶à×ÊÔ´Çë¼Ócqhdy2016",
    "genre": "Ò»ÊÖ×ÊÔ´¹Ù·½Í¬²½¸üÐÂÇë¼ÓÎ¢ÐÅ£ºcqhdy2016",
    "recordingDate": "2017-01-01T00:00:00",
    "title": "1 - 李靖难成仙 下山当总兵"
  },
```
to 
```json
  "meta": {
    "album": "",
    "artist": "",
    "comment": "",
    "genre": "",
    "recordingDate": "2017-01-01T00:00:00",
    "title": "1 - 李靖难成仙 下山当总兵"
  },
```

then I tried to apply the changes back to the `.mp3`, 
```bash
$ tone tag . --auto-import ToneJson
Tagging 193 files, continue? [y/n] (y): y
── unchanged: /mnt/g/audio-test/超极封神榜/1 - 李靖难成仙 下山当总兵.mp3 ──────────────────────────────────────────────────────────────────────
```
No luck.

I also tried 
```shell
$ tone tag .  --auto-import Ffmetadata
Tagging 193 files, continue? [y/n] (y): y
Error: Object reference not set to an instance of an object.
```
no luck either.


BTW, I am testing on WSL(Ubuntu) on Windows 10

**Comment by sandreas** [comment]
> then I picked one of the tone json files. changed(cleared) some fields

Ah, I see. Clearing fields is special... I can't simply remove fields that have an empty value or are not set, because in this case it would remove EVERY tag, that is not in the export file which is mostly not the wanted behaviour in most cases. 

`atldotnet` treats `""` (empty string) same as `null` in most cases and cannot differ between UNSET value and SET="". I already discussed this with the maintainer and was working on a fix/workaround, but I did not have the time to finish it.

To remove fields, use e.g. `--meta-remove-property=album`. Maybe I need to reinvestigate this for this feature.

**Comment by sandreas** [comment]
Phew, you give me a lot to investigate. 
According to [this](https://github.com/sandreas/DotnetLibAudioMetadata/blob/13fd4927c92dab74ab13aca09aa510f1095e9b69/Sandreas.AudioMetadata/AudioMetadata/IMetadataExtensions.cs#L435) line of code, it should work like expected.

I update the featurette in my first comment - so I have to check, what is wrong here. Sorry, might take a while.

**Comment by rockonedege** [comment]
Never mind. Thanks for the help indeed.

If you need me to test, just let me know... and take your time. I am in no hurry.
